### PR TITLE
etcd_svc: tolerate unready endpoints

### PR DESCRIPTION
**Pull request by hongchaodeng**
fix https://github.com/kubernetes-incubator/bootkube/issues/599

### CODE REVIEW

**Comment by hongchaodeng** [comment]
@xiang90 @aaronlevy @diegs

**Comment by xiang90** [comment]
looks good to me.

@aaronlevy where is the best place to track the alpha feature we use?

**Comment by diegs** [comment]
IIUC We should not be using alpha features by default in bootkube.

**Comment by fxposter** [comment]
@diegs But this annotation doesn't break anything, isn't it? It is not possible to achieve this behavior for previous versions of kubernetes, so you lose nothing.

**Comment by hongchaodeng** [comment]
@diegs 
So this has some offline discussion with @aaronlevy and @xiang90 . Let me share it on github too.

First of all, we have seen many problems with unready (or not "Running") endpoints in service:
- https://github.com/kubernetes-incubator/bootkube/issues/599
- https://github.com/coreos/tectonic-installer/issues/1266

We would need this feature from k8s to make the whole thing more robust. Otherwise, it's very fragile and only work under **good assumptions**.

Second, this is not something new in k8s world. There has been long-discussed issues (https://github.com/kubernetes/kubernetes/issues/39363, https://github.com/kubernetes/kubernetes/issues/25283) for how to handle stateful workloads on k8s. This is something that upstream is moving towards (even slowly) for stateful workloads. Many statefulset applications are using it.

Third, as mentioned above, this doesn't break anything. Further more, self-hosted-etcd is also alpha feature.

**Comment by diegs** [comment]
Thanks for the additional context. It also looks like despite being an alpha feature it's enabled all the time (https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go) so it seems ok.

Can you add a brief YAML comment to the template saying a little bit about why we need this alpha feature (and also maybe mention that it has been around a long time & is always enabled)? That way people who are deriving their own configs from the templates will understand why it's helpful.

**Comment by diegs** [comment]
Though it looks like the migration to self-hosted etcd failed in the e2e test:

```
Migrating to self-hosted etcd cluster...
I0705 17:59:32.627115     997 log.go:19] warning: ignoring ServerName for user-provided CA for backwards compatibility is deprecated
I0705 17:59:42.630782     997 migrate.go:65] created etcd cluster TPR
Error: services "etcd-service" not found
Tearing down temporary bootstrap control plane...
Error: services "etcd-service" not found
Error: services "etcd-service" not found
services "etcd-service" not found
```

**Comment by aaronlevy** [comment]
@xiang90 maybe start a `Documentation/alpha-features.md` document that tracks what alpha features are being used, and why.

Ideally we aren't using alpha features -- but we had discussed the risk in enabling this and it seemed low enough to at least try and exercise it with self-hosted etcd. 

- the feature is already enabled by default in the controller-manager (just opt-in with annotation)
- it's used already by / relied upon by statefulsets
- we still consider self-hosted etcd to be in alpha itself.

My main concern would be people upgrading their clusters in the future and this feature essentially being graduated to an actual field, which means the annotation would just silently stop working unless they knew to update to the new fields. Generally covered by release notes, but just something to consider (and why I'd like to have an alpha feature tracking list).

**Comment by aaronlevy** [comment]
Also agree with @diegs that a brief note directly in the YAML would be helpful too

**Comment by xiang90** [comment]
> @xiang90 maybe start a Documentation/alpha-features.md document that tracks what alpha features are being used, and why.

sounds good.

**Comment by hongchaodeng** [comment]
comments added

**Comment by hongchaodeng** [comment]
> maybe start a Documentation/alpha-features.md document

We will write this thing in etcd-operator repo.
@aaronlevy Can you write that in bootkube repo?

File: Documentation/alpha-features.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Alpha Features
+
```
Comment (aaronlevy) [inline Documentation/alpha-features.md]: Add a quick description of this document. Something like:

> Tracking document for alpha features that bootkube installed clusters may make use of.
>
> We track these alpha features as their behavior may change or be deprecated between Kubernetes versions. Therefore clusters that use these features need to keep track of any potential changes in upstream releases. See the upstream [api versioning documentation](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#alpha-beta-and-stable-versions) for more information.

File: Documentation/alpha-features.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+### TolerateUnreadyEndpointsAnnotation
+
```
Comment (aaronlevy) [inline Documentation/alpha-features.md]: Add a note about what component uses this. Something like:

> Used by the etcd service object when self-hosted etcd cluster is enabled.

File: Documentation/alpha-features.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+This alpha annotation will retain the endpoints even if the etcd pod isn't ready.
```
Comment (aaronlevy) [inline Documentation/alpha-features.md]: Can you also add a link to issue that discusses why this is needed?

**Comment by hongchaodeng** [comment]
@aaronlevy 
fixed

**Comment by hongchaodeng** [comment]
coreosbot run e2e

**Comment by hongchaodeng** [comment]
coreosbot run e2e

**Comment by xiang90** [comment]
coreosbot run e2e

**Comment by hongchaodeng** [comment]
coreosbot run e2e

**Comment by hongchaodeng** [comment]
coreosbot run e2e

**Comment by hongchaodeng** [comment]
coreosbot run e2e

**Comment by xiang90** [comment]
coreosbot run e2e

**Comment by xiang90** [comment]
coreosbot run e2e

**Comment by hongchaodeng** [comment]
coreosbot run e2e

**Comment by hongchaodeng** [comment]
coreosbot run e2e

**Comment by diegs** [comment]
@hongchaodeng rebasing may help with the tests.

**Comment by hongchaodeng** [comment]
pass after rebase

**Pull request merged**